src/expr-parser/src/parser.rs

```diff
@@ -0,0 +1,1483 @@
+    // Define a Parser that constructs a (read-only) parsing context `ctx` and
+    // delegates to `relation::parse_expr` by passing a `ctx` as a shared ref.
+    let parser = move |input: ParseStream| {
+        let ctx = Ctx { catalog };
+        relation::parse_expr(&ctx, input)
+    };
+    // Since the syn lexer doesn't parse comments, we replace all `// {`
+    // occurrences in the input string with `:: {`.
+    let s = s.replace("// {", ":: {");
+    // Call the parser with the given input string.
+    let mut expr = parser.parse_str(&s).map_err(|err| {
+        let (line, column) = (err.span().start().line, err.span().start().column);
+        format!("parse error at {line}:{column}:\n{err}\n")
+    })?;
+    // Fix the types of the local let bindings of the parsed expression in a
+    // post-processing pass.
+    relation::fix_types(&mut expr, &mut relation::FixTypesCtx::default())?;
+    // Return the parsed, post-processed expression.
+    Ok(expr)
+}
+
+/// Builds a source definition from a string.
+pub fn try_parse_def(catalog: &TestCatalog, s: &str) -> Result<Def, String> {
+    // Define a Parser that constructs a (read-only) parsing context `ctx` and
+    // delegates to `relation::parse_expr` by passing a `ctx` as a shared ref.
+    let parser = move |input: ParseStream| {
+        let ctx = Ctx { catalog };
+        def::parse_def(&ctx, input)
+    };
+    // Call the parser with the given input string.
+    let def = parser.parse_str(s).map_err(|err| {
+        let (line, column) = (err.span().start().line, err.span().start().column);
+        format!("parse error at {line}:{column}:\n{err}\n")
+    })?;
+    // Return the parsed, post-processed expression.
+    Ok(def)
+}
+
+/// Support for parsing [mz_expr::MirRelationExpr].
+mod relation {
+    use std::collections::BTreeMap;
+
+    use mz_expr::{Id, JoinImplementation, LocalId, MirRelationExpr};
+    use mz_repr::{Diff, RelationType, Row, ScalarType};
+
+    use super::*;
+
+    type Result = syn::Result<MirRelationExpr>;
+
+    pub fn parse_expr(ctx: CtxRef, input: ParseStream) -> Result {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(kw::Constant) {
+            parse_constant(ctx, input)
+        } else if lookahead.peek(kw::Get) {
+            parse_get(ctx, input)
+        } else if lookahead.peek(kw::Return) {
+            parse_let(ctx, input)
+        } else if lookahead.peek(kw::Project) {
+            parse_project(ctx, input)
+        } else if lookahead.peek(kw::Map) {
+            parse_map(ctx, input)
+        } else if lookahead.peek(kw::FlatMap) {
+            parse_flat_map(ctx, input)
+        } else if lookahead.peek(kw::Filter) {
+            parse_filter(ctx, input)
+        } else if lookahead.peek(kw::CrossJoin) {
+            parse_cross_join(ctx, input)
+        } else if lookahead.peek(kw::Join) || lookahead.peek(kw::CrossJoin) {
```

Comment:
The `|| lookahead.peek(kw::CrossJoin)` is not needed here, because it will match in the previous `if`, right?